File: claude/.claude/skills/code-sec/fixtures/vuln-app/node/safe.js
// Safe Express fixture — the correct variant of every vuln in vuln.js.
//
// GREEN bed: each rule that FIRES on the matching vuln.js sink must stay SILENT
// here. Same routes, same entry points, sinks sanitized. Ground truth: ../MANIFEST.md.

const express = require("express");
const { execFile } = require("child_process");
const fs = require("fs");
const path = require("path");
const axios = require("axios");

const app = express();
app.use(express.json());
const db = { query: (sql, params, cb) => cb(null, []) }; // stand-in driver

const ACCOUNTS = { 1: { owner: "alice", balance: 42 }, 2: { owner: "bob", balance: 99 } };
const ALLOWED_HOSTS = new Set(["api.internal.example"]);
const DATA_DIR = "/var/data";

app.get("/order", (req, res) => {
  // SAFE CWE-89 — parameterized query, value bound not interpolated.
  db.query("SELECT * FROM orders WHERE id = ?", [req.query.id], (err, rows) => res.json(rows)); // SAFE:CWE-89
});

app.get("/ping", (req, res) => {
  // SAFE CWE-78 — execFile with an argument vector, no shell.
  execFile("ping", ["-c", "1", req.query.host], (err, stdout) => res.send(stdout)); // SAFE:CWE-78
});

app.get("/fetch", (req, res) => {
  // SAFE CWE-918 — host allowlist before the outbound request.
  const host = new URL(req.query.url).hostname;
  if (!ALLOWED_HOSTS.has(host)) return res.status(403).end();
  return axios.get(req.query.url).then((r) => res.send(r.data)); // SAFE:CWE-918
});

app.get("/download", (req, res) => {
  // SAFE CWE-22 — basename strips traversal; join stays under DATA_DIR.
  const name = path.basename(req.query.name);
  fs.readFile(path.join(DATA_DIR, name), "utf8", (err, data) => res.send(data)); // SAFE:CWE-22
});

app.post("/import", (req, res) => {
  // SAFE CWE-502 — JSON.parse builds only plain data, never code.
  const obj = JSON.parse(req.body.data); // SAFE:CWE-502
  res.json(obj);
});

app.get("/accounts/:id", (req, res) => {
  // SAFE CWE-639 — ownership check against the session user before returning.
  const acct = ACCOUNTS[req.params.id];
  if (!acct || acct.owner !== req.session.user) return res.status(404).end();
  return res.json(acct); // SAFE:CWE-639
});

app.listen(8081, "127.0.0.1"); // local bind

File: claude/.claude/skills/code-sec/fixtures/vuln-app/python/safe.py
"""Safe Flask fixture — the correct variant of every vuln in vuln.py.

GREEN bed: each rule that FIRES on the matching vuln.py sink must stay SILENT
here. Same routes, same entry points, sinks sanitized. Ground truth: ../MANIFEST.md.
"""

import os
import subprocess
from urllib.parse import urlparse

import requests
import yaml
from flask import Flask, abort, request, session

app = Flask(__name__)

ORDERS = {"1": {"owner": "alice", "total": 42}, "2": {"owner": "bob", "total": 99}}
ALLOWED_HOSTS = {"api.internal.example"}
DATA_DIR = "/var/data"


def _current_user():
    return session.get("user")


@app.route("/order")
def get_order():
    # SAFE CWE-89 — parameterized query, no interpolation.
    order_id = request.args["id"]
    cur = _db().cursor()
    cur.execute("SELECT * FROM orders WHERE id = ?", (order_id,))  # SAFE:CWE-89
    return str(cur.fetchall())


@app.route("/ping")
def ping():
    # SAFE CWE-78 — argument vector, no shell, host validated.
    host = request.args["host"]
    if not host.replace(".", "").isalnum():
        abort(400)
    out = subprocess.run(["ping", "-c", "1", host], capture_output=True)  # SAFE:CWE-78
    return out.stdout


@app.route("/fetch")
def fetch():
    # SAFE CWE-918 — host allowlist before the outbound request.
    url = request.args["url"]
    if urlparse(url).hostname not in ALLOWED_HOSTS:
        abort(403)
    return requests.get(url, timeout=5).text  # SAFE:CWE-918


@app.route("/download")
def download():
    # SAFE CWE-22 — basename strips traversal; join stays under DATA_DIR.
    name = os.path.basename(request.args["name"])
    with open(os.path.join(DATA_DIR, name)) as fh:  # SAFE:CWE-22
        return fh.read()


@app.route("/config", methods=["POST"])
def load_config():
    # SAFE CWE-502 — SafeLoader refuses arbitrary object construction.
    return str(yaml.safe_load(request.data))  # SAFE:CWE-502


@app.route("/orders/<order_id>")
def read_order(order_id):
    # SAFE CWE-639 — ownership check against the session user before returning.
    order = ORDERS.get(order_id)
    if order is None or order["owner"] != _current_user():
        abort(404)
    return str(order)  # SAFE:CWE-639


@app.route("/admin/delete")
def admin_delete():
    # SAFE CWE-287 — authorization gate before the privileged action.
    if _current_user() != "admin":
        abort(403)
    target = request.args["user"]
    return f"deleted {target}"  # SAFE:CWE-287


def _db():
    import sqlite3

    return sqlite3.connect(":memory:", check_same_thread=False)


if __name__ == "__main__":
    app.run(host="127.0.0.1", port=8081)  # local bind
